### PR TITLE
Kaldi 5.2 (the batch-norm + convolutional PR).

### PR DESCRIPTION
**Pull request by danpovey**
I'll make the home of this in the official Kaldi repo, as I'll be asking people to make PRs to it.

### CODE REVIEW

**Comment by francisr** [comment]
When do you think you'll be able to merge this into master?

**Comment by danpovey** [comment]
Maybe a week or two more.

**Comment by danpovey** [comment]
@jtrmal, I wonder if you have any comment on this..

A recent commit of this PR removes the support from train_rnn.py, train_dnn.py and so on, for discriminative pretraining, i.e. it no longer supports setups that have layer1.config, layer2.config, and so on.  This is the kind of change that there will never really be a good time to make it, but it's a simplification that's helpful for the health of the code.

I searched for the scripts that are invalidated by this change, and the following is what I got.
They include some BABEL scripts, but I don't know how important they are. 

```
./egs/ami/s5/local/chain/run_blstm_ami_5.sh
./egs/ami/s5b/local/nnet3/run_lstm.sh
./egs/aspire/s5/local/nnet3/run_lstm.sh
./egs/aspire/s5/local/chain/tuning/run_blstm_7b.sh
./egs/aspire/s5/local/chain/tuning/run_blstm_asp_1.sh
./egs/babel/s5d/local/chain/run_blstm.sh
./egs/babel/s5d/local/chain/run_blstm_bab1.sh
./egs/babel/s5d/local/chain/run_blstm_bab2.sh
./egs/babel/s5d/local/chain/run_blstm_bab3.sh
./egs/babel/s5d/local/chain/run_blstm_bab4.sh
./egs/babel/s5d/local/chain/run_blstm_bab5.sh
./egs/babel/s5d/local/chain/run_blstm_xconfig.sh
./egs/babel/s5d/local/nnet3/run_lstm.sh
./egs/babel/s5d/local/nnet3/run_lstm_realigned.sh
./egs/fisher_swbd/s5/local/chain/run_blstm_6h.sh
./egs/fisher_swbd/s5/local/nnet3/run_lstm.sh
./egs/gale_arabic/s5b/local/nnet3/tuning/run_lstm_1a.sh
./egs/sprakbanken/s5/local/nnet3/run_lstm.sh
./egs/swbd/s5c/local/chain/tuning/run_blstm_6h.sh
./egs/swbd/s5c/local/chain/tuning/run_blstm_6i.sh
./egs/swbd/s5c/local/chain/tuning/run_lstm_6h.sh
./egs/swbd/s5c/local/chain/tuning/run_lstm_6i.sh
./egs/swbd/s5c/local/chain/tuning/run_lstm_d.sh
./egs/swbd/s5c/local/nnet3/run_lstm.sh
./egs/tedlium/s5_r2/local/nnet3/tuning/run_lstm_1a.sh
./egs/multi_en/s5/local/chain/run_blstm_6h.sh
```

**Comment by jtrmal** [comment]
I didn't manage to fully migrate the chain (nor nnet3 scripts) to
xconfigs, but looking at the files,

```

./egs/babel/s5d/local/chain/run_blstm.sh
./egs/babel/s5d/local/chain/run_blstm_bab1.sh
./egs/babel/s5d/local/chain/run_blstm_bab2.sh
./egs/babel/s5d/local/chain/run_blstm_bab3.sh
./egs/babel/s5d/local/chain/run_blstm_bab4.sh
./egs/babel/s5d/local/chain/run_blstm_bab5.sh
./egs/babel/s5d/local/chain/run_blstm_xconfig.sh
./egs/babel/s5d/local/nnet3/run_lstm.sh
./egs/babel/s5d/local/nnet3/run_lstm_realigned.sh

```

they are not really part of the main pipeline (we were using chain
tdnn and nnet3 blstm).

So I'd say these can actually go.

y.

**Comment by danpovey** [comment]
OK, so I'll remove those scripts in the kaldi_52 branch.

**Comment by jtrmal** [comment]
BTW, what exactly is wrong about this file so that it got removed? It is using xconfig, so I'm a little confused

https://github.com/kaldi-asr/kaldi/blob/6e53b883d25706d2817b64f28d4bfd11f4833f4f/egs/babel/s5d/local/chain/run_blstm_xconfig.sh

y.

**Comment by danpovey** [comment]
Oh sorry, I'll un-delete it.  I was using grep to figure out the list, and
I think it may have had an old script in a comment.

**Comment by danpovey** [comment]
I've got results for the first time on CIFAR-10 in an nnet3-based setup now.  It's taking about half an hour to train (with 1-2 GPUs, for 30 epochs; no augmentation).  Test-set accuracy (reported below as 'validation') is about 76%, and training-set accuracy 98%.  It's not tuned at all yet.

a04:v1: grep Overall  exp/cnn1a_cifar10/log/train.?.?0.log 
grep: exp/cnn1a_cifar10/log/train.?.?0.log: No such file or directory
a04:v1: grep accuracy exp/cnn1a_cifar10/log/compute_prob_valid.?0.log
exp/cnn1a_cifar10/log/compute_prob_valid.10.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.6855 per frame, over 10000 frames.
exp/cnn1a_cifar10/log/compute_prob_valid.20.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.7424 per frame, over 10000 frames.
exp/cnn1a_cifar10/log/compute_prob_valid.30.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.7432 per frame, over 10000 frames.
exp/cnn1a_cifar10/log/compute_prob_valid.40.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.7615 per frame, over 10000 frames.
exp/cnn1a_cifar10/log/compute_prob_valid.50.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.7639 per frame, over 10000 frames.


grep accuracy exp/cnn1a_cifar10/log/compute_prob_train.?0.log
exp/cnn1a_cifar10/log/compute_prob_train.10.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.7358 per frame, over 5000 frames.
exp/cnn1a_cifar10/log/compute_prob_train.20.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.8448 per frame, over 5000 frames.
exp/cnn1a_cifar10/log/compute_prob_train.30.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.8838 per frame, over 5000 frames.
exp/cnn1a_cifar10/log/compute_prob_train.40.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.9652 per frame, over 5000 frames.
exp/cnn1a_cifar10/log/compute_prob_train.50.log:LOG (nnet3-compute-prob[5.1.113~1-05d11]:PrintTotalStats():nnet-diagnostics.cc:146) Overall accuracy for 'output'\
 is 0.9816 per frame, over 5000 frames.

**Comment by danpovey** [comment]
I'm commenting here on github so more people can see the discussion.

It looks like there are two issues with the batch-norm, and I want to fix them both before we merge to master.
First- and this is something @GaofengCheng pointed out- there is something in the original BatchNorm paper that I neglected to implement, with two trainable parameters, beta and gamma, per dimension.  It seems that this is something that people do use, and it should probably be implemented.  For instance, @hhadian mentioned that in Keras they do implement that part.

Secondly, it seems that the way we are getting the batch-norm stats to use in test time is not working well at all.  We are getting them from the final training iteration, but it seems that the stats change too fast even within an iteration, for this to be a good strategy.  If I recompute the stats without training the model (or use --test-mode=false when computing the train or valid probs), the probs are quite a bit better, at least with CNN-based models.  The interaction with the final 'combine' stage is also not very good.  So this will have to be changed also.  I'll try to work on it myself because this is a limiting factor for a bunch of different things.

**Comment by GaofengCheng** [comment]
FYI @danpovey
 https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md

**Comment by danpovey** [comment]
Thanks, Gaofeng!  That link was extremely useful.
A couple of takeaways from their results:
  - What worked best for them was to skip the scale and shift (the beta and
gamma of the paper) and put the batch-norm after the nonlinearity.  As it
happens that was exactly the way I imagined it was supposed to be done when
I implemented it, so maybe we *don't* have to implement the scale and shift.
  - They have some interesting experiments on initialization.  In
particular, 'orthonormal' initialization works well for them (ensure the
parameter matrices have orthonormal rows.. see this blog post
http://hjweide.github.io/orthogonal-initialization-in-convolutional-layers).
That could easily be done in our setup.

Something else, not about the blog post: in my current good-working CNN
setup after I messed with the learning rate of the 1st convolutional layer
to stop it training too fast, the problem with the batch-norm stats getting
out of date disappeared.  But I'm still planning to look into the best way
to get the batch-norm stats, and to do the final 'combine' stage with
batch-norm (i.e. test-mode vs. not).

File: egs/wsj/s5/local/chain/tuning/run_tdnn_1b.sh [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+    --feat.online-ivector-dir=$train_ivector_dir \
+    --feat.cmvn-opts="--norm-means=false --norm-vars=false" \
+    --chain.xent-regularize $xent_regularize \
+    --chain.leaky-hmm-coefficient=0.1 \
+    --chain.l2-regularize=0.00005 \
+    --chain.apply-deriv-weights=false \
+    --chain.lm-opts="--num-extra-lm-states=2000" \
+    --trainer.srand=$srand \
+    --trainer.max-param-change=2.0 \
+    --trainer.num-epochs=4 \
```
Comment (GaofengCheng) [inline egs/wsj/s5/local/chain/tuning/run_tdnn_1b.sh]: @danpovey I find here you use 4 epochs for WSJ, should I try 4 or 7 epochs or both.. on AMI ?

**Comment by danpovey** [comment]
Your choice.  I tried, on WSJ, using more epochs (6, I think), but it
didn't help.

**Comment by danpovey** [comment]
I am planning to merge this today, so officially master will point to kaldi 5.2 and this will be revision 5.2.0.  I'll wait for the checks to finish.
All PRs from that point on should be to master, the kaldi_52 branch will stop moving and will eventually be deleted.
